Source/WebCore/platform/mediastream/gstreamer/GStreamerVideoEncoder.cpp

```diff
@@ -462,21 +546,63 @@ static void webkit_webrtc_video_encoder_class_init(WebKitWebrtcVideoEncoderClass
+            switch (mode) {
+            case CONSTANT_BITRATE_MODE:
+                gst_util_set_object_arg(G_OBJECT(encoder), "pass", "cbr");
+                break;
+            case VARIABLE_BITRATE_MODE:
+                gst_util_set_object_arg(G_OBJECT(encoder), "pass", "pass1");
+                break;
+            };
+        }, [](GstElement* encoder, LatencyMode mode) {
+            switch (mode) {
+            case REALTIME_LATENCY_MODE:
+                gst_util_set_object_arg(G_OBJECT(encoder), "tune", "zerolatency");
+                gst_util_set_object_arg(G_OBJECT(encoder), "speed-preset", "ultrafast");
+                break;
+            case QUALITY_LATENCY_MODE:
+                g_object_set(encoder, "tune", 0, nullptr);
+                gst_util_set_object_arg(G_OBJECT(encoder), "speed-preset", "None");
+                gst_util_set_object_arg(G_OBJECT(encoder), "pass", "qual");
+                break;
+            };
+        });
     Encoders::registerEncoder(OpenH264, "openh264enc", "h264parse", "video/x-h264",
-        "video/x-h264,alignment=au,stream-format=byte-stream",
-        [](WebKitWebrtcVideoEncoder* self) {
+        "video/x-h264,alignment=au,stream-format=byte-stream", [](WebKitWebrtcVideoEncoder* self) {
             g_object_set(self->priv->parser.get(), "config-interval", 1, nullptr);
             g_object_set(self->priv->outputCapsFilter.get(), "caps", self->priv->encodedCaps.get(), nullptr);
-        }, "bitrate", setBitrateBitPerSec, "gop-size");
+        }, "bitrate", setBitrateBitPerSec, "gop-size", [](GstElement*, BitrateMode) {
+            notImplemented();
+        }, [](GstElement*, LatencyMode) {
+            notImplemented();
+        });
     Encoders::registerEncoder(Vp8, "vp8enc", nullptr, "video/x-vp8", nullptr,
         [](WebKitWebrtcVideoEncoder* self) {
-            gst_preset_load_preset(GST_PRESET(self->priv->encoder.get()), "Profile Realtime");
-        }, "target-bitrate", setBitrateBitPerSec, "keyframe-max-dist");
+            gst_util_set_object_arg(G_OBJECT(self->priv->encoder.get()), "keyframe-mode", "disabled");
+        }, "target-bitrate", setBitrateBitPerSec, "keyframe-max-dist", [](GstElement* encoder, BitrateMode mode) {
+            switch (mode) {
+            case CONSTANT_BITRATE_MODE:
+                gst_util_set_object_arg(G_OBJECT(encoder), "end-usage", "cbr");
+                break;
+            case VARIABLE_BITRATE_MODE:
+                gst_util_set_object_arg(G_OBJECT(encoder), "end-usage", "vbr");
+                break;
+            };
+        }, [](GstElement* encoder, LatencyMode mode) {
+            switch (mode) {
+            case REALTIME_LATENCY_MODE:
+                gst_preset_load_preset(GST_PRESET(encoder), "Profile Realtime");
+                break;
+            case QUALITY_LATENCY_MODE:
+                g_object_set(encoder, "threads", 4, "cpu-used", 4, "deadline", 0, "lag-in-frames", 25, nullptr);
```

Comment:
Where do these 4 threads and 4 cpus come from? Why is it cabled here? Shouldn't this be a parameter somehow? (I mean here and everywhere this is used) Maybe even just a `#define`?

Reply:
I think that dates back from original code Thibault wrote? idk.

Reply:
I can add a define for now and let us revisit this later on...